### PR TITLE
Login attempts repository

### PR DESCRIPTION
**Pull request by James-Tepper**

### CODE REVIEW

File: server/repositories/login_attempts.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+READ_PARAMS = """\
+    login_attempt_id,
+    successful,
+    ip_address,
+    user_agent,
+    created_at
+"""
+
+
+class Login_Attempt(TypedDict):
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
class LoginAttempt(TypedDict):
```

File: server/repositories/login_attempts.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    user_agent: str
+    created_at: datetime
+
+
+async def create(
+    login_attempt_id: int,
+    successful: bool,
+    ip_address: str,
+    user_agent: str,
+) -> Login_Attempt:
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
) -> LoginAttempt:
```

File: server/repositories/login_attempts.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        values={
+            "login_attempt_id": login_attempt_id,
+            "successful": successful,
+            "ip_address": ip_address,
+            "user_agent": user_agent,
+        },
+    )
+
+    assert login_attempt is not None
+    return cast(Login_Attempt, login_attempt)
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
    return cast(LoginAttempt, login_attempt)
```

File: server/repositories/login_attempts.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        query=f"""\
+            SELECT {READ_PARAMS}
+            FROM login_attempts
+            WHERE login_attempt_id = :login_attempt_id
+        """,
+        values={
+            "login_attempt_id": login_attempt_id,
+        },
+    )
+    return cast(Login_Attempt, login_attempt) if login_attempt is not None else None
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
    return cast(LoginAttempt, login_attempt) if login_attempt is not None else None
```

File: server/repositories/login_attempts.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+
+async def fetch_many(
+    login_attempt_id: int,
+    page: int | None = None,
+    page_size: int | None = None,
+) -> list[Login_Attempt] | None:
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
) -> list[LoginAttempt] | None:
```

File: server/repositories/login_attempts.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    if page is not None and page_size is not None:
+        query += f"""\
+            LIMIT :page_size
+            OFFSET :offset
+        """
+        values["limit"] = page_size
+        values["offset"] = (page - 1) * page_size
+
+    login_attempt = await clients.database.fetch_all(query, values)
+    return [cast(Login_Attempt, login_attempt) for attempt in login_attempt]
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
    return [cast(LoginAttempt, login_attempt) for attempt in login_attempt]
```

File: server/repositories/login_attempts.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            "ip_address": ip_address,
+            "user_agent": user_agent,
+        },
+    )
+
+    assert login_attempt is not None
+    return cast(Login_Attempt, login_attempt)
+
+
+async def fetch_one(login_attempt_id: int) -> Login_Attempt | None:
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
async def fetch_one(login_attempt_id: int) -> LoginAttempt | None:
```

File: server/repositories/login_attempts.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        """,
+        values={
+            "login_attempt_id": login_attempt_id,
+        },
+    )
+    return cast(LoginAttempt, login_attempt) if login_attempt is not None else None
+
+
+async def fetch_many(
+    login_attempt_id: int,
```
Comment (cmyui) [inline server/repositories/login_attempts.py]: a given `login_attempt_id` will only ever return a single result, or none, so it doesn't make sense for a fetch_many. only fetch_one.

for this, the filter parameters we want are the `successful`, `ip_address`, and `user_agent`

**Review by cmyui — CHANGES_REQUESTED** [review]
requested minor changes - rest looks good!

File: server/repositories/login_attempts.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    return cast(LoginAttempt, login_attempt) if login_attempt is not None else None
+
+
+async def fetch_many(
+    successful: bool | None = None,
+    ip_address: str | None = None,
+    user_agent: str | None = None,
+    page: int | None = None,
+    page_size: int | None = None,
+) -> list[LoginAttempt] | None:
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
) -> list[LoginAttempt]:
```

**Review by metal-face — APPROVED** [review]

File: server/repositories/login_attempts.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    if page is not None and page_size is not None:
+        query += f"""\
+            LIMIT :page_size
+            OFFSET :offset
+        """
+        values["limit"] = page_size
+        values["offset"] = (page - 1) * page_size
+
+    login_attempt = await clients.database.fetch_all(query, values)
+    return [cast(LoginAttempt, attempt) for attempt in login_attempt]
```
Comment (cmyui) [inline server/repositories/login_attempts.py]:
```suggestion
    login_attempts = await clients.database.fetch_all(query, values)
    return [cast(LoginAttempt, attempt) for attempt in login_attempts]
```

**Review by cmyui — APPROVED** [review]

**Review by cmyui — APPROVED** [review]

**Pull request merged**